File: app/services/anchors.py
from __future__ import annotations

import html
import uuid
from typing import Any, Iterable

from app.editor.models import normalize_block, normalize_blocks


def anchor_name(block: dict[str, Any]) -> str:
    if block.get("type") != "anchor":
        return ""
    data = block.get("data", {})
    return str(data.get("text") or data.get("name") or "").strip()


def anchor_display_name(block: dict[str, Any]) -> str:
    if block.get("type") != "anchor":
        return ""
    data = block.get("data", {})
    return str(data.get("display_name") or anchor_name(block)).strip()


def anchor_target_id(block: dict[str, Any]) -> str | None:
    if block.get("type") != "anchor":
        return None
    value = block.get("data", {}).get("target_block_id")
    return str(value) if value else None


def anchor_targets(
    blocks: Iterable[dict[str, Any]],
    *,
    exclude_ids: Iterable[str] = (),
) -> list[dict[str, Any]]:
    excluded = {str(value) for value in exclude_ids}
    return [
        block
        for block in blocks
        if block.get("type") != "anchor" and str(block.get("id")) not in excluded
    ]


def linked_anchors(
    blocks: Iterable[dict[str, Any]],
    target_block_id: str,
) -> list[dict[str, Any]]:
    target = str(target_block_id)
    return [block for block in blocks if anchor_target_id(block) == target]


def new_anchor_data(
    display_name: str,
    target_block_id: str,
    *,
    existing_names: Iterable[str] = (),
) -> dict[str, Any]:
    label = " ".join(str(display_name).split()).strip()[:64]
    used = {str(value) for value in existing_names}
    name = ""
    while not name or name in used:
        name = f"anchor_{uuid.uuid4().hex[:10]}"
    return {
        "text": name,
        "display_name": label,
        "target_block_id": str(target_block_id),
        "html": f'<a name="{html.escape(name, quote=True)}"></a>',
    }


def set_anchor_display_name(block: dict[str, Any], display_name: str) -> bool:
    if block.get("type") != "anchor":
        return False
    label = " ".join(str(display_name).split()).strip()[:64]
    if not label:
        return False
    block.setdefault("data", {})["display_name"] = label
    return True


def set_anchor_target(
    blocks: list[dict[str, Any]],
    anchor_id: str,
    target_block_id: str,
) -> bool:
    target = next(
        (
            block
            for block in blocks
            if str(block.get("id")) == str(target_block_id)
            and block.get("type") != "anchor"
        ),
        None,
    )
    anchor = next(
        (
            block
            for block in blocks
            if str(block.get("id")) == str(anchor_id)
            and block.get("type") == "anchor"
        ),
        None,
    )
    if anchor is None or target is None:
        return False
    anchor.setdefault("data", {})["target_block_id"] = str(target["id"])
    align_linked_anchors(blocks)
    return True


def retarget_linked_anchors(
    blocks: list[dict[str, Any]],
    old_target_id: str,
    new_target_id: str,
) -> bool:
    target = next(
        (
            block
            for block in blocks
            if str(block.get("id")) == str(new_target_id)
            and block.get("type") != "anchor"
        ),
        None,
    )
    anchors = linked_anchors(blocks, old_target_id)
    if target is None or not anchors:
        return False
    for anchor in anchors:
        anchor.setdefault("data", {})["target_block_id"] = str(target["id"])
    align_linked_anchors(blocks)
    return True


def align_linked_anchors(blocks: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Keep every linked anchor immediately before its target block.

    Imported/legacy anchors without a target remain exactly where the user put
    them, so old pages keep their existing behavior.
    """
    normalize_blocks(blocks)
    target_ids = {
        str(block.get("id"))
        for block in blocks
        if block.get("type") != "anchor"
    }
    grouped: dict[str, list[dict[str, Any]]] = {}
    linked_ids: set[str] = set()
    for block in blocks:
        target_id = anchor_target_id(block)
        if target_id and target_id in target_ids:
            grouped.setdefault(target_id, []).append(block)
            linked_ids.add(str(block.get("id")))

    ordered: list[dict[str, Any]] = []
    for block in blocks:
        if str(block.get("id")) in linked_ids:
            continue
        block_id = str(block.get("id"))
        ordered.extend(grouped.get(block_id, ()))
        ordered.append(block)
    blocks[:] = ordered
    for index, block in enumerate(blocks):
        normalize_block(block, position=index)
    return blocks


__all__ = [
    "align_linked_anchors",
    "anchor_display_name",
    "anchor_name",
    "anchor_target_id",
    "anchor_targets",
    "linked_anchors",
    "new_anchor_data",
    "retarget_linked_anchors",
    "set_anchor_display_name",
    "set_anchor_target",
]

File: lib/editor/anchors.js
import { normalizeBlock, normalizeBlocks } from 'lib/editor/models';

export function anchorName(block) {
  if (block?.type !== 'anchor') return '';
  return String(block?.data?.text ?? block?.data?.name ?? '').trim();
}

export function anchorDisplayName(block) {
  if (block?.type !== 'anchor') return '';
  return String(block?.data?.display_name ?? anchorName(block)).trim();
}

export function anchorNavigationRichText(blocks) {
  const links = [];
  const ordered = [...(blocks ?? [])].sort((a,b) => Number(a?.position ?? 0) - Number(b?.position ?? 0));
  for (const block of ordered) {
    if (block?.type !== 'anchor') continue;
    const label = String(block?.data?.display_name ?? '').trim().replace(/\s+/g,' ');
    const name = anchorName(block);
    if (!label || !name) continue;
    if (links.length) links.push(' · ');
    links.push({ type:'anchor_link', text:label, anchor_name:name });
  }
  if (!links.length) return null;
  return links.length === 1 ? links[0] : links;
}

export function anchorTargetId(block) {
  if (block?.type !== 'anchor') return null;
  const value = block?.data?.target_block_id;
  return value ? String(value) : null;
}

export function anchorTargets(blocks, { excludeIds = [] } = {}) {
  const excluded = new Set([...excludeIds].map(String));
  return (blocks ?? []).filter((block) => block?.type !== 'anchor' && !excluded.has(String(block?.id)));
}

export function linkedAnchors(blocks, targetBlockId) {
  const target = String(targetBlockId);
  return (blocks ?? []).filter((block) => anchorTargetId(block) === target);
}

function randomAnchorName() {
  return `anchor_${(Date.now().toString(16) + Math.floor(Math.random()*1e9).toString(16)).slice(-10)}`;
}

export function newAnchorData(displayName, targetBlockId, { existingNames = [] } = {}) {
  const label = String(displayName).trim().replace(/\s+/g,' ').slice(0,64);
  const used = new Set([...existingNames].map(String));
  let name = '';
  do { name = randomAnchorName(); } while (!name || used.has(name));
  return {
    text:name,
    display_name:label,
    target_block_id:String(targetBlockId),
    html:`<a name="${name.replaceAll('"','&quot;')}"></a>`,
  };
}

export function setAnchorDisplayName(block, displayName) {
  if (block?.type !== 'anchor') return false;
  const label = String(displayName).trim().replace(/\s+/g,' ').slice(0,64);
  if (!label) return false;
  block.data ??= {};
  block.data.display_name = label;
  return true;
}

export function setAnchorTarget(blocks, anchorId, targetBlockId) {
  const target = (blocks ?? []).find((block) => String(block?.id) === String(targetBlockId) && block?.type !== 'anchor');
  const anchor = (blocks ?? []).find((block) => String(block?.id) === String(anchorId) && block?.type === 'anchor');
  if (!target || !anchor) return false;
  anchor.data ??= {};
  anchor.data.target_block_id = String(target.id);
  alignLinkedAnchors(blocks);
  return true;
}

export function retargetLinkedAnchors(blocks, oldTargetId, newTargetId) {
  const target = (blocks ?? []).find((block) => String(block?.id) === String(newTargetId) && block?.type !== 'anchor');
  const anchors = linkedAnchors(blocks, oldTargetId);
  if (!target || !anchors.length) return false;
  for (const anchor of anchors) {
    anchor.data ??= {};
    anchor.data.target_block_id = String(target.id);
  }
  alignLinkedAnchors(blocks);
  return true;
}

export function alignLinkedAnchors(blocks) {
  normalizeBlocks(blocks);
  const targetIds = new Set(blocks.filter((b)=>b?.type!=='anchor').map((b)=>String(b.id)));
  const grouped = new Map();
  const linkedIds = new Set();
  for (const block of blocks) {
    const targetId = anchorTargetId(block);
    if (targetId && targetIds.has(targetId)) {
      if (!grouped.has(targetId)) grouped.set(targetId,[]);
      grouped.get(targetId).push(block);
      linkedIds.add(String(block.id));
    }
  }
  const ordered = [];
  for (const block of blocks) {
    if (linkedIds.has(String(block.id))) continue;
    const blockId = String(block.id);
    ordered.push(...(grouped.get(blockId) ?? []), block);
  }
  blocks.splice(0,blocks.length,...ordered);
  blocks.forEach((block,index)=>normalizeBlock(block,{position:index}));
  return blocks;
}
